src/transaction.rs

```diff
@@ -1,78 +1,360 @@
-use crate::nullifier::Nullifier;
-use crate::circuit::validity_predicate::ValidityPredicate;
-use crate::el_gamal::EncryptedNote;
-use crate::{
-    action::Action, add_to_tree, is_in_tree, note::Note, serializable_to_vec, CircuitParameters,
+use crate::action::Action;
+use crate::circuit::action_circuit::ActionCircuit;
+use crate::circuit::blinding_circuit::BlindingCircuit;
+use crate::circuit::circuit_parameters::CircuitParameters;
+use crate::circuit::validity_predicate::{ValidityPredicate, NUM_NOTE};
+use crate::constant::{
+    ACTION_CIRCUIT_SIZE, ACTION_PUBLIC_INPUT_CM_INDEX, ACTION_PUBLIC_INPUT_NF_INDEX,
+    ACTION_PUBLIC_INPUT_ROOT_INDEX, BLINDING_CIRCUIT_SIZE,
 };
-use ark_ec::twisted_edwards_extended::GroupAffine as TEGroupAffine;
-use rs_merkle::{algorithms::Blake2s, MerkleTree};
-
-pub struct Transaction<'a, CP: CircuitParameters> {
-    //max: usize, // the maximum number of actions/notes for a transaction
-    actions: Vec<Action<CP>>,
-    spent_notes: Vec<(Note<CP>, Nullifier<CP>)>,
-    created_notes: Vec<(Note<CP>, EncryptedNote<CP::InnerCurve>)>,
-    vps: &'a Vec<ValidityPredicate<CP>>,
+use crate::error::TaigaError;
+use crate::vp_description::ValidityPredicateDescription;
+use plonk_core::circuit::Circuit;
+use plonk_core::circuit::{verify_proof, VerifierData};
+use plonk_core::proof_system::{pi::PublicInputs, Proof, VerifierKey};
+use rand::RngCore;
+
+pub const NUM_TX_SLICE: usize = NUM_NOTE;
```

Comment:
Each transaction slice corresponds to spending one note and creating one note?

Reply:
If yes, could we consider a note consumption without creating a new one? Maybe it does not make sense ahah

Reply:
It's an action slice with one input note and one output note to generate an action proof.
Both input note and output note can be dummy(zero, or invalid note) so that we can construct an action without creating a new one.